addon_service/oauth1/utils.py

```diff
@@ -0,0 +1,148 @@
+from base64 import b64encode
+from hashlib import sha1
+from hmac import HMAC
+from http import (
+    HTTPMethod,
+    HTTPStatus,
+)
+from secrets import token_urlsafe
+from time import time
+from urllib.parse import (
+    parse_qs,
+    quote_plus,
+)
+
+from aiohttp import ContentTypeError
+
+from addon_service.common.aiohttp_session import get_singleton_client_session
+from addon_toolkit.credentials import OAuth1TokenCredentials
+from addon_toolkit.iri_utils import iri_with_query
+
+
+async def get_request_token(
+    request_token_url: str,
+    oauth_consumer_key: str,
+    oauth_consumer_secret: str,
+) -> tuple[OAuth1TokenCredentials, dict]:
+    """
+    Obtaining unauthorised request token needed to construct Authorization url
+    https://oauth.net/core/1.0a/#auth_step1
+    """
+    signed_headers = construct_signed_headers(
+        request_token_url, oauth_consumer_key, oauth_consumer_secret
+    )
+    return await _get_token(request_token_url, signed_headers)
+
+
+async def get_access_token(
+    access_token_url: str,
+    oauth_consumer_key: str,
+    oauth_consumer_secret: str,
+    oauth_token: str,
+    oauth_token_secret: str,
+    oauth_verifier: str,
+) -> tuple[OAuth1TokenCredentials, dict]:
+    """
+    Getting final access token needed to access protected resources from Service provider
+
+    """
+    signed_headers = construct_signed_headers(
+        access_token_url,
+        oauth_consumer_key,
+        oauth_consumer_secret,
+        oauth_token=oauth_token,
+        oauth_verifier=oauth_verifier,
+        oauth_token_secret=oauth_token_secret,
+    )
+    return await _get_token(access_token_url, signed_headers)
+
+
+async def _get_token(
+    request_token_url, signed_headers
+) -> tuple[OAuth1TokenCredentials, dict]:
+    _client = await get_singleton_client_session()
+    async with _client.post(
+        request_token_url, headers=signed_headers
+    ) as _token_response:
+        if not HTTPStatus(_token_response.status).is_success:
+            raise RuntimeError(await _token_response.text())
+        try:
+            return OAuth1TokenCredentials.from_dict(await _token_response.json())
+        except ContentTypeError:
+            raw_result = parse_qs(await _token_response.text())
+            result_dict = {key: value[0] for key, value in raw_result.items() if value}
+            return OAuth1TokenCredentials.from_dict(result_dict)
+
+
+def _construct_params(params_to_encode: dict) -> str:
+    return ",".join(
+        [f'{key}="{value}"' for key, value in sorted(params_to_encode.items())]
```

Comment:
```suggestion
        f'{key}="{value}"' for key, value in sorted(params_to_encode.items())
```
nit: don't need the intermediate list